kie-server-parent/kie-server-controller/kie-server-controller-openshift/src/main/java/org/kie/server/controller/openshift/storage/ServerTemplateOpenShiftStorage.java

```diff
@@ -0,0 +1,237 @@
+import org.kie.server.controller.api.model.spec.ServerTemplateKey;
+import org.kie.server.controller.api.storage.KieServerTemplateStorage;
+import org.kie.server.services.openshift.impl.storage.cloud.KieServerStateOpenShiftRepository;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+public class ServerTemplateOpenShiftStorage implements KieServerTemplateStorage {
+
+    private static final Logger logger = LoggerFactory.getLogger(ServerTemplateOpenShiftStorage.class);
+    private static final String SERVER_TEMPLATE_KEYS_CACHE_KEY = "server.template.keys.cache.key";
+    private static final String SERVER_TEMPLATE_CACHE_KEY = "server.template.cache.key";
+    private static final long DEFAULT_CACHE_TTL = 60000L; // 60 seconds
+
+    static SimpleInMemoryCache storageCache = new SimpleInMemoryCache();
+
+    KieServerStateOpenShiftRepository repo;
+    long cacheTTL = DEFAULT_CACHE_TTL;
+
+    public ServerTemplateOpenShiftStorage() {
+        this(new KieServerStateOpenShiftRepository());
+    }
+
+    public ServerTemplateOpenShiftStorage(KieServerStateOpenShiftRepository repo) {
+        this.repo = repo;
+    }
+
+    @Override
+    public ServerTemplate store(ServerTemplate serverTemplate) {
+        if (exists(serverTemplate.getId())) {
+            throw new IllegalArgumentException("Server template with id " + serverTemplate.getId() + " is already stored");
+        }
+        repo.create(ServerTemplateConverter.toState(serverTemplate));
+        storageCache.clear();
+        return ServerTemplateConverter.fromState(repo.load(serverTemplate.getId()));
+    }
+
+    @Override
+    public List<ServerTemplateKey> loadKeys() {
+        @SuppressWarnings("unchecked")
+        List<ServerTemplateKey> result = (List<ServerTemplateKey>) storageCache.get(SERVER_TEMPLATE_KEYS_CACHE_KEY);
+        if (result == null) {
+            result = repo.retrieveAllKieServerIds()
+                         .stream()
+                         .map(id -> new ServerTemplateKey(id, id))
+                         .collect(Collectors.toList());
+            storageCache.add(SERVER_TEMPLATE_KEYS_CACHE_KEY, result, cacheTTL);
+        }
+        return result;
+    }
+
+    @Override
+    public List<ServerTemplate> load() {
+        @SuppressWarnings("unchecked")
+        List<ServerTemplate> result = (List<ServerTemplate>) storageCache.get(SERVER_TEMPLATE_CACHE_KEY);
+        if (result == null) {
+            result = repo.retrieveAllKieServerStates()
+                         .stream()
+                         .map(ServerTemplateConverter::fromState)
+                         .collect(Collectors.toList());
+            storageCache.add(SERVER_TEMPLATE_CACHE_KEY, result, cacheTTL);
+        }
+        return result;
+    }
+
+    @Override
+    public ServerTemplate load(String identifier) {
+        ServerTemplate template = null;
+        try {
+            template = ServerTemplateConverter.fromState(repo.load(identifier));
```

Comment:
Originally, we thought that caching is good for bulk loading methods such as load(), loadKeys(), and it need to tolerate staled data for short period of time. For loading individual KieServerState/ServerTemplate, it would be better not to employing caching.  What do you think?

Reply:
@rhtevan Wouldn't this approach bring inconsistencies, when bulk load returns different results than load of single template?

Reply:
Initial investigating shows that the data used by UI rending server template detail is exclusively retrieved by single load method rather than the bulk load method. Therefore, the risk of inconsistency and yielding negative impact is low. At the moment, we have agreed on keeping the caching strategy implement as simple as necessary.  

https://github.com/kiegroup/kie-wb-common/blob/master/kie-wb-common-screens/kie-wb-common-server-ui/kie-wb-common-server-ui-client/src/main/java/org/kie/workbench/common/screens/server/management/client/ServerManagementBrowserPresenter.java#L167